### ISSUE TITLE
Shared mutable state

### ISSUE THREAD

**Issue by dgopstein** [SLP-team/SLP-Core#3]
Currently, the three Runner classes (LexerRunner, ModelRunner, VocabularyRunner) all rely on static fields to configure how they train models. Configuring these classes can be difficult to sequence when one is dealing with multiple models concurrently. My intuition is this it might benefit encapsulation to move some of the static state either inside of an individual object, or as parameters to the static methods that already exist.

For example, I had created two models which I wanted to compare against each other, however in between training/evaluating each model I learned I needed to re-initialize each runner for the model I was testing. Before I learned this, however, I was getting unexpected behavior, since training one model could re-configure the runner used by the other model.

My original error:
```java
// Compare two different model configurations (by varying the per-line parameter)
LexerRunner.setPerLine(false);
ModelRunner.learn(model1, train1);

LexerRunner.setPerLine(true);
ModelRunner.learn(model2, train1);

// Later compare each model using a different training set
ModelRunner.learn(model1, train2); // per-line ought to be set to 'false' here, but it's not
ModelRunner.learn(model2, train2);
```

Potential solution - embed configuration parameters in object, and pass that to training
```java
Lexer lexer1 = new JavaLexer() {{perLine = false;}};
Lexer lexer2 = new JavaLexer() {{perLine = true;}};

ModelRunner.learn(lexer1, model1, train1)
ModelRunner.learn(lexer2, model2, train1)

ModelRunner.learn(lexer1, model1, train2)
ModelRunner.learn(lexer2, model2, train2)
```

This type of approach has a few advantages:
 * Less verbose (since the Runners do not need to be reconfigured in between each usage)
 * Parallelizable (since parameters aren't shared between models, they can be run interchangeably)
  * Explicit (since Runner configurations are specified statically, not dynamically)

On the negative side, a change like this might be a pretty large-scale refactor.

**Comment by VHellendoorn** [comment]
Thanks for the suggestion. I do agree that it is time to get rid of the static state. It wasn't there in v0.1, and was only added in v0.2 because it was easier to write. Parallelization is probably most hindered by the current setup, and mixing models that use different vocabularies/lexers is almost impossible as well for the same reason. Verbosity is an interesting aspect: I envisioned that most configurations would be frozen after setup in typical use, but the scenario you are describing certainly defies that. I wonder how often such issues come up.

As for implementation, the issue of having these be non-static is where the configuration needs to come from: we either have to pass a vocabulary, lexer and model anywhere where the model is needed (_e.g._ to `ModelRunner.learn`), or every `Model` needs to come with its own `Lexer` and `Vocabulary` field (which doesn't seem right). Would you lean towards the former?
I am also not a fan of the configuration parameter pattern (which some call an [anti-pattern](https://blog.jooq.org/2014/12/08/dont-be-clever-the-double-curly-braces-anti-pattern/) because it makes a new class for every invocation). I'd rather write this using something like a `LexerConfiguration` class that can be passed to a `Lexer` (with some standards set by the respective lexers).

 I'll make a branch with these changes once we agree on a design; it shouldn't be terribly difficult, but it will break some of the interfaces (_esp._ interfaces like `Lexer` may have to become abstract classes to support carrying configurations).

**Comment by dgopstein** [comment]
I'm not sure my verbosity argument much matters, it was just a thing that occurred to me as I was typing. Relatedly, I've only been using this library for a couple days, so my comments should be regarded mostly as first impressions.

I think your suggestion of using a `LexerConfiguration` object is a good one. I was just using the `{{...}}` for a simple example. Perhaps a good balance is the [Builder Pattern](http://minborgsjavapot.blogspot.com/2014/08/creating-objects-using-builder-pattern.html) recommended in Effective Java.

I definitely agree that it feels wrong for a Model to know/care about the data-flow before it, and so it shouldn't possess information about a Lexer/Vocabulary. Perhaps the annoyance of having to specify all three parameters to e.g. `ModelRunner` could be mitigated by providing a meta-object to lump them all together, e.g.: `new ModelContext(lexer, vocabulary, model)`. Then this single object is then the only parameter that's needed in a lot of scenarios. I do wonder if that approach then makes it harder to operate on the individual components, such as mixing a global/local model together.

**Comment by dgopstein** [comment]
Upon further consideration, maybe my idea of inventing a `ModelContext` is a half-step towards a larger idea of converting `ModelRunner` from static to instance methods. If ModelRunner were instantiable, you could effectively do:

```java
ModelRunner mr = new ModelRunner(lexer, vocab, model);

mr.learn(file);
```

I'm not sure whether or not this is a good idea, but it is a logical extension of my previous idea.

**Comment by VHellendoorn** [comment]
I like the idea of making the Runners instantiable, perhaps leaving a small amount of static functionality in them. This way, a `ModelRunner` can reasonably ask for a `LexerRunner` instance even though a `Model` remains oblivious to the `Lexer` that provided its tokens. I'll give this a shot and see what complications arise; will post the link to the branch when I'm done

**Comment by VHellendoorn** [comment]
I just pushed the first draft of the new Runners to dev, commit 0c2bd92. In this update, the `LexerRunner` and `ModelRunner` are no longer static. Some design decisions that we can discuss:
- The `ModelRunner` now takes a `LexerRunner` as argument, and the `LexerRunner` requires a `Vocabulary`. The `Model` and `Lexer` interfaces remain as they were, agnostic of other stages in the pipeline.
- The `VocabularyRunner` remains as was, since it really just provided some util functions for creating/loading/writing `Vocabulary`s. The `Vocabulary` itself is now no longer globally static.

Some issues with the current design that could be addressed before merging to master:
- The NestedModel now requires a `ModelRunner` as a parameter. This is because it is really a set of models that changes shape over time, so it needs access to this functionality. However, this does make it a circular dependency, which should be mitigated. I considered making it a ModelRunner sub-class, but that doesn't quite pan out yet.
- The `ModelRunner` is too big. This was already a problem, but at least as a static util class it's defensible. I'm thinking of refactoring it into a more decent structure before making it main-line.
- I haven't yet converted `ModelRunner` and `LexerRunner` into builder-pattern classes, but I think it makes sense to do so.

Let me know what you think and if you experience any strange issues. Note that this update will break some interfaces, so please refer to the `example` classes again for insights into fixing those issues. Mostly it means instantiating the ModelRunner and LexerRunner.

**Comment by dgopstein** [comment]
That was really quick turn-around on the diff! Thanks :)

I'm reading through the changes trying to respond to your comment above. If you're only interested in high-level feedback, then I think this forum is perfect.

On the other hand, if you're interested in line-level code comments as well it might make sense to turn your commits into a pull-request even if you're ultimately the person who'd accept the request. The UI for code-review is just easier that way.

Either way is good for me, let me know!

**Comment by VHellendoorn** [comment]
Let's stick to a high level, and I'll PR it into master when it seems good enough. Let me know if you have any thoughts on the design decisions and issues I mentioned; it's pretty easy to change these things for me.

**Comment by dgopstein** [comment]
The updated structure of `ModelRunner` and `LexerRunner`, and lack of update for `VocabularyRunner` all make sense to me.

To a zoom in a little on `ModelRunner`, I have two small questions:
* Will an individual `ModelRunner` object be useful for multiple models, or will each new model want its own specific model runner? A 1-to-1 relationship might imply that `ModelRunner`'s constructor should take a model as an argument. This would have the benefit simplifying the signatures of many methods in the class, and explicitly linking a `Lexer` to a `Model`.
* Is there any reason someone would want to instantiate `new ModelRunner()` with no arguments? I accidentally used that constructor when porting my code over and got confused why my `LexerRunner`’s regex was being ignored.

Regarding the mentioned issues:
* It appears that the two places that `NestedModel` needs to run itself from are the constructor and the `notify()` mechanism. Are both necessary?
    * Is there a reason why  `NestedModel`s (as opposed to other models) must train their local component upon construction, rather than in an explicit call to `.learn()`?
    * I don’t fully understand NestedModel's `.notify` method, but it seems to learn some data, cache the resultant model, then unlearn the data. Would it be a reasonable idea/possible to do most of this from inside `modelRunner.learn`/`.forget`/`.predict` instead? Maybe the model also shares some responsibility in this task, but perhaps that functionality is more specific?
* While the Builder-pattern idea might be theoretically nice for `Runner`s, I think the current usage of setters is perfectly reasonable.
    * On the other hand I think the `Model`s might benefit from a Builder. Currently the `NGramModel` constructors dispatch on different types of numbers (e.g. `new JMModel(1)` is very different from `new JMModel(1.0)`). Using builders here would (a) differentiate the arguments better (b) reduce redundancy of these constructors being redefined in each subclass of `NGramModel`.
* TLDR - `ModelRunner` is little long, but IMO that’s mostly superficial and isn’t necessarily a high-level-design type problem.
    * I definitely see that `ModelRunner` has a lot of lines. And normally lots of lines implies lots of functionality, which might imply that functionality is not cohesive/specific enough. In this case, I think a lot of the size actually comes from reimplementations of similar functionality across different data structures, e.g. the same method written twice, once for `ArrayList` and once for `Stream`. I think this is mostly good news because it means that the class isn’t philosophically bloated, just awkward to write in Java. Moreover, as your comment `TODO, invoke next method if possible` implies, I think the code-duplication is definitely reduceable. As far as the actual design/perview of the class it seems to be 90% focused on taking models and using them to learn/forget/predict - which is sounds pretty ideal for a `ModelRunner`.

**Comment by VHellendoorn** [comment]
Thanks for your feedback. It makes sense to me to give each ModelRunner its own Model. I also removed the ModelRunner-with-no-argument constructor to avoid confusion.

The nested model needs some rethinking, so I'm not pushing my code just yet, but I should have a solution for that soon. I'm toying with making it a subclass to ModelRunner instead, since its functionality (including in `notify`) is primarily about setting up a hierarchy of models that morphs with the file being modeled.

I like the idea of a builder pattern for Models, though it seems pretty independent from the current issue. Perhaps you can make a PR with that change somewhere down the line?

**Comment by VHellendoorn** [comment]
I'm opening #6 for this issue (and for #5, although I am not sure if that is resolved by the change I proposed there). Unfortunately, it is fairly tangled because of the major interface overhaul; changes are documented in the description and I would appreciate if you could just see if the code works intuitively for you. diff-specific comments would be a bit too time-consuming on the PR. Let me know if you have any suggestions and I'll attach the commits to that PR; when I merge it, the code moves to v0.4.

**Comment by VHellendoorn** [comment]
Resolved in #6
